File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  let i = 0
  let j = 0
  let maxLength = 1
  let addedChars = new Set();
  if(s == "")
      return 0
  if(s.length == 1)
      return 1
  while (j < s.length) {
      if(!addedChars.has(s.charAt(j))) {
          addedChars.add(s.charAt(j))
          j++
          maxLength = Math.max(maxLength, addedChars.size)
      }
      else {
          addedChars.delete(s.charAt(i))
            i++ 
      }
      
  }
    
    return maxLength
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int start = 0;
        int end = 0;
        int l = 0;
        
        // Map<String, String> occurrenceMap = new HashMap<>();
        HashSet<Character> set = new HashSet<>();
        
        if(s.length() <= 1)
            return s.length();
        
        while(end < s.length()) {
            if(set.contains(s.charAt(end))) {
                // repeated char, increment start
                set.remove(s.charAt(start));
                start++;
            } else {
                // char is not repeated, include it
                // alse, add it to hashmap
                set.add(s.charAt(end));
                l = Math.max(l, end - start + 1);
                end++;
            }
        }
        
        return l;
    }
}